### PR TITLE
Adding the Service and Election Rumors

### PR DESCRIPTION
**Pull request by adamhjk**
gif-keyboard-11192703965568213462

* Adds a Service rumor, for declaring that this member has a running
  service
* Adds an Election rumor, which does leader election amongst all the
  members who have the running service.
* Adds a RumorStore, which abstracts storing the various rumors
* Adds tests for all of the above

Signed-off-by: Adam Jacob <adam@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/butterfly/src/server/mod.rs [hunk 1]
```diff
@@ -281,36 +317,223 @@ impl Server {
...
+                  total_population);
+            return false;
+        }
+        if total_population % 2 == 0 {
+            warn!("This census has an even population. If half the membership fails, quorum will \
+                   never be met, and no leader will be elected. Add another instance to the \
+                   service group!");
+        }
+
+        let alive_pop = alive_population as f32;
```
Comment (stevendanna) [inline components/butterfly/src/server/mod.rs]: What are the types of `alive_population` and `total_population`?  If they are an integer type, I'm wondering if the following might be more clear:

```
alive_population > ((total_population / 2) + 1)
```

as it avoid the casts and round()s.

Comment (adamhjk) [inline components/butterfly/src/server/mod.rs]: Done! You've unlocked my weakness - arithmetic.

File: .travis.yml [hunk 2]
```diff
@@ -66,7 +66,7 @@ matrix:
   - language: rust
     env:
       - COMPONENTS=lib LIBSODIUM_PREFIX=$HOME/pkgs/libsodium/1.0.8 LIBARCHIVE_PREFIX=$HOME/pkgs/libarchive/3.2.0 LIBZMQ_PREFIX=$HOME/pkgs/zeromq/4.1.4 PKG_CONFIG_PATH="$PKG_CONFIG_PATH:$LIBARCHIVE_PREFIX/lib/pkgconfig:$LIBSODIUM_PREFIX/lib/pkgconfig:$LIBZMQ_PREFIX/lib/pkgconfig" LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$LIBARCHIVE_PREFIX/lib:$LIBSODIUM_PREFIX/lib:$LIBZMQ_PREFIX/lib" LIBRARY_PATH="$LIBRARY_PATH:$LIBZMQ_PREFIX/lib"
-      - AFFECTED_DIRS="components/builder-dbcache|components/builder-protocol|components/common|components/core|components/builder-depot-client|components/http-client|components/net"
+      - AFFECTED_DIRS="components/builder-dbcache|components/builder-protocol|components/common|components/core|components/builder-depot-client|components/http-client|components/net|components/butterfly"
```
Comment (reset) [inline .travis.yml]: It seems like we miss one of these monday details every time we add a project ;)

File: components/butterfly/src/rumor.rs [hunk 3]
```diff
@@ -26,23 +26,162 @@
...
+use error::{Result, Error};
 
 /// The description of a `RumorKey`.
-#[derive(Clone, Debug, Hash, PartialEq, PartialOrd, Eq)]
+///
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
 pub struct RumorKey {
-    pub kind: String,
-    pub key: UuidSimple,
+    pub kind: Rumor_Type,
```
Comment (reset) [inline components/butterfly/src/rumor.rs]: :+1:

File: components/butterfly/src/election.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        self.get_term() == other.get_term()
+    }
+}
+
+impl Rumor for Election {
+    /// Updates this election based on the contents of another election.
+    fn merge(&mut self, mut other: Election) -> bool {
+        if *self == other {
+            // If we are the same object, just return false
+            // println!("Equal: {:?} {:?}", self, other);
```
Comment (reset) [inline components/butterfly/src/election.rs]: Do you want to leave this in as `debug!()` or remove it completely?

File: components/butterfly/src/election.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            // floor
+            false
+        } else if self.get_term() > other.get_term() {
+            // If the rumor we got has a term that's lower than ours, keep sharing our rumor no
+            // matter what term they are on.
+            true
+        } else if self.get_suitability() > other.get_suitability() {
+            // If we are more suitable than the other side, we want to steal
+            // the other sides votes, and keep sharing.
+            // println!("Self suitable: {:?} {:?}", self, other);
```
Comment (reset) [inline components/butterfly/src/election.rs]: `debug()!` here as well? There's a few below in this module if you want to take a peek at those, too :)

File: components/butterfly/src/rumor.rs [hunk 6]
```diff
@@ -26,23 +26,162 @@
...
+            update_counter: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+}
+
+impl<T: Rumor + Clone> RumorStore<T> {
+    /// Create a new RumorStore for the given type. Allows you to initialize the counter to a
+    /// pre-set value. Useful mainly in testing.
+    pub fn new(counter: usize) -> RumorStore<T> {
+        RumorStore { update_counter: Arc::new(AtomicUsize::new(counter)), ..Default::default() }
```
Comment (reset) [inline components/butterfly/src/rumor.rs]: Love this syntax, I didn't know about it until reading here `{ thing: T, ..Default::default() }`

File: components/butterfly/src/rumor.rs [hunk 7]
```diff
@@ -132,66 +271,188 @@ impl RumorList {
...
+            false
+        }
+
+        fn write_to_bytes(&self) -> Result<Vec<u8>> {
+            Ok(Vec::from(format!("{}-{}", self.id, self.key).as_bytes()))
         }
     }
 
+    #[derive(Clone, Debug)]
     pub struct TrumpRumor {
```
Comment (reset) [inline components/butterfly/src/rumor.rs]: I didn't even notice this solid gold before

File: components/butterfly/src/server/mod.rs [hunk 8]
```diff
@@ -281,36 +317,220 @@ impl Server {
...
+        let total_population = self.service_store.len_for_key(key);
+        let alive_population = electorate.len();
+
+        if total_population < 3 {
+            info!("Quorum size: {}/3 - election cannot complete",
+                  total_population);
+            return false;
+        }
+
+        if total_population % 2 == 0 {
```
Comment (reset) [inline components/butterfly/src/server/mod.rs]: So we must *always* have an odd number of supervisors?

File: components/butterfly/src/server/push.rs [hunk 9]
```diff
@@ -163,51 +160,76 @@ impl PushWorker {
...
+                        }
+                    }
                 }
-            };
-            let rumor_as_bytes = match send_rumor.write_to_bytes() {
-                Ok(bytes) => bytes,
-                Err(e) => {
-                    println!("Could not write our own rumor to bytes; abandoning \
+                ProtoRumor_Type::Service => {
+                    // trace_it!(GOSSIP: &self.server, TraceKind::SendRumor, member.get_id(), &send_rumor);
```
Comment (reset) [inline components/butterfly/src/server/push.rs]: Did you mean to disable tracing here and below on 193?

**Comment by reset** [comment]
gif-keyboard-12933566464025171069

@adamhjk the testing continues to impress me. Great work here!

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/171485961).

File: components/butterfly/src/rumor.rs (hunk 7 above)
Comment (adamhjk) [inline components/butterfly/src/rumor.rs]: gif-keyboard-8653824803627861390

File: components/butterfly/src/election.rs (hunk 4 above)
Comment (adamhjk) [inline components/butterfly/src/election.rs]: I'll remove it

File: components/butterfly/src/server/mod.rs (hunk 8 above)
Comment (adamhjk) [inline components/butterfly/src/server/mod.rs]: If you don't want a warning, yes. If you have an even number of services in a group, and you want an election, you run the risk of both sides having quorum, which means you won't actually ever be able to complete an election.

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/171485961).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

File: components/butterfly/src/server/push.rs (hunk 9 above)
Comment (adamhjk) [inline components/butterfly/src/server/push.rs]: Only for a second, while I figure out how to re-enable it. :)

**Pull request merged**